# HPE Container Platform API series  - Lab 2
## Deploying cloud-native stateless applications and non-cloud-native stateful applications, programmatically on Kubernetes cluster managed by HPE Container Platform.


**Requirements:**
- HPE Container Platform deployment with a managed Kubernetes cluster
- IP address or FQDN of the HPE Container Platform's controller host and Gateway host
- a Kubernetes tenant user account   

**Utilities:**   
- cURL  
- Jupyter Notebook server with bash kernel installed
- kubectl

**Definitions:**
- *HPE Container Platform* is an enterprise-grade container platform designed to deploy both cloud-native and non-cloud-native applications whether on-premises, at the edge, in multiple public clouds, or in a hybrid model. This makes the HPE Container Platform ideal for helping application developers and data scientists accelerate their application development and deployment on **containers**, on-demand through a self-service portal and a RESTful API that surfaces programmable access. To learn more about HPE Container Platform visit the [HPE DEV portal](https://developer.hpe.com/platform/hpe-container-platform/home) and check out the blog articles.

- *tenant:* A tenant is a group of users created by the platform administrator. A tenant can represent for example, an office location, a business unit, an organization, a project, an application. A tenant is allocated a quota of resources (CPU/GPU, RAM, Storage, App Store images, Kubernetes cluster) by the platform administrator. All the resources used by a tenant are not shared with other tenants. A tenant user is granted the role of member or admin for the tenant.


**Lab workflow:**

In this lab part:

1. As tenant user you will first establish a valid login session with HPE CP using HPE CP REST API.

2. You will then fetch the Kubeconfig file for your tenant working context using HPE CP REST API.

3. Finally, using a K8s API client such as **kubectl** you will deploy a simple cloud-native *stateless*, microservices applications, and a non-cloud-native distributed *stateful* AI/analytics **KubeDirector** application on **Kubernetes** cluster managed by HPE Container Platform. 


>Note: kubectl has been installed in the JupyterHub server by the IT administrator.

- *KubeDirector:* also known as Kubernetes Director, is a key component of HPE Container Platform. KubeDirector is an open source project initiated by HPE (BlueData) that enables the running of non-cloud-native stateful analytics workloads on Kubernetes. In HPE Container Platform, these applications generally refer to a distributed, single-node or multi-node application **virtual cluster**. Each application virtual cluster node runs as **a container** in the HPE Container Platform.

- *Stateless application:* Stateless application is an application which does not require persistence of data and application state.

- *Stateful application:* A stateful application typically requires persistence of certain mountpoints across application cluster nodes rescheduling, restarts, upgrades, rollbacks. A stateful application can also need persistence of network identity (i.e.: hostname). 

#### Initialize the environment.

**IMPORTANT: Before running the next cells, please make sure to check/adjust the environment variables below according to your Student ID, tenant username and password.**

In [1]:
#
# environment variables to be adjusted/verified by the student
#
student="denis" # your Jupyter Notebook student Identifier (i.e.: student<xx>)
username="demouser" # your HPE CP tenant login credentials - username 
password="password" # your HPE CP tenant login credentials - password
#
# fixed environment variables setup by the HPE CP lab administrator
#
controller_endpoint="gateway1.hpedevlab.net:8080"
controller_host="gateway1.hpedevlab.net"
tenantname="K8sHackTenant"
k8sClusterId="1"  #this is the K8s Cluster Id provided by the HPE CP admisnistrator and assigned to your K8s tenant.
helloWorldApp="hello-world-app.yaml" # the application manifest you will deploy in this lab
tensorFlowApp="tensorflow-notebook-config-cluster.yaml" # the kubedirector application cluster configuration TensorFlow

#### Authenticate as Tenant user in the specified tenant:

In [2]:
sessionlocation=$(curl -k -i -s --request POST "https://${controller_endpoint}/api/v2/session" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--data-raw '{
"name": "'"$username"'",
"password": "'"$password"'",
"tenant_name": "'"$tenantname"'"
}' | grep Location | awk '{print $2}' | tr -d '\r') #we remove any cr that might exist
echo "This is your session location: " $sessionlocation
SessionId=$(echo $sessionlocation | cut -d'/' -f 5) # extract sessionId for later, for logout
echo "This is your session_Id:" $SessionId

This is your session location:  /api/v2/session/44d9f376-b8a1-43eb-bf38-cb71451b94df
This is your session_Id: 44d9f376-b8a1-43eb-bf38-cb71451b94df


#### Make a quick check to ensure you can make REST API calls within your tenant working context:
Here you will fetch information about your session you have just established.

In [3]:
curl -k -s --request GET "https://${controller_endpoint}/api/v2/session/${SessionId}" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' | jq  #using jq to pretty print the JSON reponse of the API call 

{
  "_links": {
    "self": {
      "href": "/api/v2/session/44d9f376-b8a1-43eb-bf38-cb71451b94df"
    },
    "all_sessions": {
      "href": "/api/v2/session"
    }
  },
  "user": "/api/v1/user/60",
  "user_name": "demouser",
  "tenant": "/api/v1/tenant/5",
  "tenant_name": "K8sHackTenant",
  "role": "/api/v1/role/3",
  "role_name": "Member",
  "expiry": "2020-5-10 18:36:00",
  "expiry_time": 1589128560,
  "is_site_admin_view": false,
  "is_cluster_superuser": false
}


#### Get the Kubeconfig file for your tenant working context
The next step in deploying a containerized application in Kubernetes clusters managed by HPE Container Platform, is to get the kubeconfig file for your tenant working context. 

The HPE CP REST API call below allows you to get the **kubeconfig file** used to access the Kubernetes cluster for your tenant user account based on your assigned role (tenant member), with the same result as if you had downloaded it from the HPE CP UI.

In [4]:
curl -k -s --request GET "https://${controller_endpoint}/api/v2/k8skubeconfig" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' > ${student}_kubeconfig

In [5]:
cat ${student}_kubeconfig

apiVersion: v1
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://gateway1.etc.fr.comm.hpecorp.net:9500
  name: k8scluster1
contexts:
- context:
    cluster: k8scluster1
    user: HPECP-demouser
    namespace: k8shacktenant
  name: k8scluster1-K8sHackTenant-demouser
current-context: k8scluster1-K8sHackTenant-demouser
kind: Config
preferences: {}
users:
- name: HPECP-demouser
  user:
    exec:
      command: kubectl
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - hpecp
      - authenticate
      - gateway1.etc.fr.comm.hpecorp.net:8080
      - --hpecp-user=demouser
      - --hpecp-token=/api/v2/session/44d9f376-b8a1-43eb-bf38-cb71451b94df
      - --hpecp-token-expiry=1589128560
      - --force-reauth=false
      - --insecure-skip-tls-verify=true


> Notice the kubeconfig file includes the valid token (session location) for your current session you previously established.

##### Define the Kubeconfig file as a shell environment variable
The kubectl command-line tool (a K8s API client) uses kubeconfig file to communicate with the Kube API server of a Kubernetes cluster. We specify here the path of the kubeconfig file by setting the KUBECONFIG environment variable:

In [6]:
#define the Kubeconfig file as a shell environment variable to tell kubectl where to look for the kubeconfig file
export KUBECONFIG="${student}_kubeconfig"
#echo $KUBECONFIG

##### Check your working tenant context

In [7]:
kubectl config view

apiVersion: v1
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://gateway1.etc.fr.comm.hpecorp.net:9500
  name: k8scluster1
contexts:
- context:
    cluster: k8scluster1
    namespace: k8shacktenant
    user: HPECP-demouser
  name: k8scluster1-K8sHackTenant-demouser
current-context: k8scluster1-K8sHackTenant-demouser
kind: Config
preferences: {}
users:
- name: HPECP-demouser
  user:
    exec:
      apiVersion: client.authentication.k8s.io/v1beta1
      args:
      - hpecp
      - authenticate
      - gateway1.etc.fr.comm.hpecorp.net:8080
      - --hpecp-user=demouser
      - --hpecp-token=/api/v2/session/44d9f376-b8a1-43eb-bf38-cb71451b94df
      - --hpecp-token-expiry=1589128560
      - --force-reauth=false
      - --insecure-skip-tls-verify=true
      command: kubectl
      env: null


You can now send Kubernetes API requests using a K8s API client such as **kubectl** to deploy enterprise workloads to the kubernetes cluster using the privileges assigned to your tenant role. 

>Note: kubectl has been installed in the JupyterHub server by the IT administrator.

**Let's see this in action!**

## Deploying a simple cloud-native stateless application

#### -1- Deploy Hello World containerized application
You will deploy a simple Hello World application on the Kubernetes cluster made available to your tenant. The K8s API call requires the kubectl operation type (create or apply) and the application manifest (YAML file that describes the attributes of the application).  

The hello-world application is a **stateless** application because it does not require persistence of data and application state. The hello-world application is a very simple application that will return `Hello Kubernetes!` when accessed. 

As you are all sharing the same tenant context and Kubernetes cluster resources, let's make sure your application deployment name will be unique among the tenant users. Here we replace the string "example" with your "username".

In [8]:
sed -i "s/example/${username}/g" $helloWorldApp
cat $helloWorldApp

apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-world-demouser
spec:
  selector:
    matchLabels:
      run: hello-demouser
  replicas: 2
  template:
    metadata:
      labels:
        run: hello-demouser
    spec:
      containers:
        - name: hello-world-demouser
          image: gcr.io/google-samples/node-hello:1.0
          ports:
            - containerPort: 8080
              protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: hello-world-service-demouser
spec:
  selector:
    run: hello-demouser
  ports:
  - name: http-hello
    protocol: TCP
    port: 8080
    targetPort: 8080
  type: NodePort


The YAML file describes the application resources to deploy such as the deployment, the POD, the container image and port, and the NodePort service used to expose the application outside the Kubernetes cluster.

In [9]:
kubectl apply -f $helloWorldApp

deployment.apps/hello-world-demouser created
service/hello-world-service-demouser created


After a minute or so, you should get the response message to your K8s API request: *Deployment and service created* on the Kubernetes cluster.   

In the next steps, you will use kubectl commands in the context of your tenant user account and get the service endpoints of your application to connect to it.

#### -2- Get the POD and Services for your deployed application:

In [10]:
kubectl get deploy,pod,service | grep ${username}

deployment.apps/hello-world-demouser   1/2     2            1           5s
pod/hello-world-demouser-54b6fcd974-j9x9b   1/1     Running             0          5s
pod/hello-world-demouser-54b6fcd974-nswmf   0/1     ContainerCreating   0          4s
service/hello-world-service-demouser   NodePort   10.96.140.179   <none>        8080:31437/TCP   5s


In [11]:
kubectl describe service hello-world-service-${username}

Name:                     hello-world-service-demouser
Namespace:                k8shacktenant
Labels:                   hpecp.hpe.com/hpecp-internal-gateway=true
Annotations:              hpecp-internal-gateway/8080: gateway1.etc.fr.comm.hpecorp.net:10002
                          kubectl.kubernetes.io/last-applied-configuration:
                            {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"name":"hello-world-service-demouser","namespace":"k8shacktenant"},"spec"...
Selector:                 run=hello-demouser
Type:                     NodePort
IP:                       10.96.140.179
Port:                     http-hello  8080/TCP
TargetPort:               8080/TCP
NodePort:                 http-hello  31437/TCP
Endpoints:                10.192.0.238:8080,10.192.1.124:8080
Session Affinity:         None
External Traffic Policy:  Cluster
Events:
  Type    Reason  Age   From         Message
  ----    ------  ----  ----         -------
  Normal  HpeCp   6s  

#### -3- Get the service endpoints
HPE Container Platform automatically maps the **NodePort Service** endpoint to the HPE Container Platform gateway (proxy) host.
Access to application services running in containers is proxied via the gateway host on a publicly-accessible IP address and a port number greater than 10000.

In [12]:
myservice="hello-world-service-"${username}
appURL=$(kubectl describe service/"${myservice}" | grep gateway1 | awk '{print $3}')
myapp_endpoint="https://${appURL}"
echo "Your application service endpoint is: "$myapp_endpoint

Your application service endpoint is: https://gateway1.etc.fr.comm.hpecorp.net:10002


#### -4- Check your application is responding:

In [13]:
curl -k -s "${myapp_endpoint}"

Hello Kubernetes!

#### -5- Delete your stateless application deployment
Delete your application deployment and services using the K8s API request below. The K8s API call requires the kubectl operation type (delete) and the application YAML manifest.
After a minute or so, you should get the message: deployment deleted and service deleted.

In [14]:
kubectl delete -f $helloWorldApp

deployment.apps "hello-world-demouser" deleted
service "hello-world-service-demouser" deleted


## Deploying a KubeDirector Application

HPE has been working within the open source Kubernetes community to add capabilities that enable the running of **stateful** analytics workloads (e.g: data-intensive, AI/ML and analytics-processing distributed applications) on Kubernetes.  The open source project is known as **Kubernetes Director** or **KubeDirector** for short.   
With KubeDirector, enterprises can deploy all their enterprise applications on a common Kubernetes framework.
KubeDirector is deployed as a custom controller (aka an operator) by default on every Kubernetes cluster managed by the HPE Container Platform.

For more information about KubeDirector, check out the HPE DEV portal [here](https://developer.hpe.com/platform/hpe-container-platform/home).

#### -1- List the installed KubeDirector Application
In the current release of HPE Container Platform, the Kubernetes clusters managed by HPE Controller platform has three (3) pre-configured KubeDirector Application types installed out of the box. A tenant user who would access the HPE Container Platform portal UI could see it:

<img src="Pictures/HPECP-KubeDirectorApp-GUI.png" height="800" width="600" align="left">

You can also get the list of KubeDirector Applications using kubectl command:

In [15]:
kubectl get kubedirectorapp

NAME                  AGE
centos7x              46d
ml-jupyter-notebook   46d
spark221e2            46d


A KubeDirector Application describes an application metadata: the service roles, the service endpoints port and port name prefix (that comes from the URL Scheme), the Docker images, the app setup packages, the cardinality (minimum number of members in a role), and if appropriate, the root file system directories (e.g.: /etc, /bin, /opt, /var, /usr) of the containers to persist beyond the life span of the containers. This means stateful applications that require to write data to their root file systems can now successfully run on Kubernetes. 

Let's inspect a couple of KubeDirector Application type definitions: 

In [16]:
kubectl describe kubedirectorapp ml-jupyter-notebook 

Name:         ml-jupyter-notebook
Namespace:    k8shacktenant
Labels:       <none>
Annotations:  token:
                execute this command to get the authentication token 'kubectl exec <pod-name> jupyter notebook list' and use this token in Jupyter noteboo...
API Version:  kubedirector.hpe.com/v1beta1
Kind:         KubeDirectorApp
Metadata:
  Creation Timestamp:  2020-03-24T08:25:15Z
  Generation:          2
  Resource Version:    2874116
  Self Link:           /apis/kubedirector.hpe.com/v1beta1/namespaces/k8shacktenant/kubedirectorapps/ml-jupyter-notebook
  UID:                 cb7a6132-f321-45d6-81f5-635f0b79e61e
Spec:
  Config:
    Role Services:
      Role ID:  controller
      Service I Ds:
        jupyter-nb
    Selected Roles:
      controller
  Config Schema Version:   7
  Default Config Package:  <nil>
  Distro ID:               bluedata/tensorflow
  Label:
    Description:  TensorFlow GPU with Jupyter notebook
    Name:         TensorFlow + Jupyter
  Roles:
    Cardinality:

In [17]:
kubectl describe kubedirectorapp spark221e2

Name:         spark221e2
Namespace:    k8shacktenant
Labels:       <none>
Annotations:  <none>
API Version:  kubedirector.hpe.com/v1beta1
Kind:         KubeDirectorApp
Metadata:
  Creation Timestamp:  2020-03-24T08:25:15Z
  Generation:          1
  Resource Version:    981021
  Self Link:           /apis/kubedirector.hpe.com/v1beta1/namespaces/k8shacktenant/kubedirectorapps/spark221e2
  UID:                 753618be-1edf-470c-bff5-385d0e76fafe
Spec:
  Config:
    Role Services:
      Role ID:  controller
      Service I Ds:
        ssh
        spark
        spark-master
        spark-worker
      Role ID:  worker
      Service I Ds:
        ssh
        spark-worker
      Role ID:  jupyter
      Service I Ds:
        ssh
        jupyter-nb
    Selected Roles:
      controller
      worker
      jupyter
  Config Schema Version:  7
  Distro ID:              bluedata/spark221e2
  Label:
    Description:  Spark 2.2.1 with Jupyter notebook
    Name:         Spark 2.2.1 + Jupyter
  Roles:
   

#### -2- Deploy the TensorFlow with Jupyter Notebook stateful application
A configuration manifest YAML file is used to create KubeDirector virtual clusters that **instantiate** the defined KubeDirector App type. The configuration file is used to describe the attributes of a given KubeDirector App.

We need to make sure the instance name of the KubeDirector Application is unique among your tenant. 

In [18]:
sed -i "s/example/${username}/g" $tensorFlowApp
cat $tensorFlowApp

---
apiVersion: "kubedirector.hpe.com/v1beta1"
kind: "KubeDirectorCluster"
metadata: 
  name: "ml-jupyter-notebook-demouser"
spec: 
  app: "ml-jupyter-notebook"
  appCatalog: "local"
  roles: 
    - 
      id: "controller"
      members: 1
      resources: 
        requests: 
          memory: "4Gi"
          cpu: "2"
          nvidia.com/gpu: "0"
        limits: 
          memory: "4Gi"
          cpu: "2"
          nvidia.com/gpu: "0"


The configuration manifest specifies the application instance name, the KubeDirectorApp type, the application service roles to deploy, the number of members (nodes) in a role and compute size, as well as the storage size if the application requires persistent storage. 

*Note: The AI/ML TensorFlow application is the easiest example to understand. In order to limit the amount of resources used in our lab HPE Container Platform environment for multiple concurrent student users, the TensorFlow KubeDirectorApp variant used here specifies a single node member and it does not specify any persistent storage.*

Let's deploy it now using the K8s API call below. The API call requires the **kubectl** operation type (create or apply) and the application YAML manifest.

In [19]:
kubectl apply -f $tensorFlowApp

kubedirectorcluster.kubedirector.hpe.com/ml-jupyter-notebook-demouser created


After a minute or so, you should get the response message to your K8s API request: *kubedirectorcluster/Your-instance-name created*.  

Once you get this message, go to the next step.

#### -3- Inspect the deployed KubeDirector application virtual cluster 
Your application virtual cluster will be represented in the Kubernetes cluster by a resource of type **KubeDirectorCluster**, with the name that was indicated inside the YAML file used to create it. **A kubeDirectorCluster resource is an instantiation of a KubeDirector App**.

In [20]:
clusterName="ml-jupyter-notebook-${username}"
kubectl get kubedirectorcluster $clusterName

NAME                           AGE
ml-jupyter-notebook-demouser   6s


After creating the instance of the KubeDirector Application, you can use kubectl command below to observe its status and the standard Kubernetes resources that compose the application virtual cluster (statefulsets, pods, services, persistent volume claim if any) and any events logged against it.

The virtual cluster status indicates its overall "state" (top-level property of the status object). It should have a value of **"configured"**. 

> Note: The first time a virtual cluster of a given KubeDirector App type is created, it may take some minutes to reach "configured" state, as the relevant Docker image must be downloaded and imported.

In [21]:
kubectl describe kubedirectorcluster $clusterName

Name:         ml-jupyter-notebook-demouser
Namespace:    k8shacktenant
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"kubedirector.hpe.com/v1beta1","kind":"KubeDirectorCluster","metadata":{"annotations":{},"name":"ml-jupyter-notebook-demouse...
API Version:  kubedirector.hpe.com/v1beta1
Kind:         KubeDirectorCluster
Metadata:
  Creation Timestamp:  2020-05-09T16:38:18Z
  Finalizers:
    kubedirector.hpe.com/cleanup
  Generation:        1
  Resource Version:  10662871
  Self Link:         /apis/kubedirector.hpe.com/v1beta1/namespaces/k8shacktenant/kubedirectorclusters/ml-jupyter-notebook-demouser
  UID:               f067b124-ef31-41ef-9bb2-63824d83dac3
Spec:
  App:          ml-jupyter-notebook
  App Catalog:  local
  Roles:
    Id:       controller
    Members:  1
    Resources:
      Limits:
        Cpu:             2
        Memory:          4Gi
        nvidia.com/gpu:  0
      Requests:
        Cpu:            

In [22]:
kubectl get all -l kubedirector.hpe.com/kdcluster=$clusterName

NAME               READY   STATUS    RESTARTS   AGE
pod/kdss-5qjtn-0   1/1     Running   0          17s

NAME                     TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
service/kdhs-mz5d5       ClusterIP   None            <none>        8888/TCP         17s
service/s-kdss-5qjtn-0   NodePort    10.96.253.106   <none>        8888:30479/TCP   17s

NAME                          READY   AGE
statefulset.apps/kdss-5qjtn   1/1     17s


Your TensorFlow KubeDirector application virtual cluster is made up of a **StatefulSet**, a **POD** (a cluster node) per service role member, a **NodePort Service** per service role member and a **headless service** for the application cluster.   

* The ClusterIP service is the headless service required by a Kubernetes StatefulSet to work. It maintains a stable POD network identity (i.e.: persistence of the hostname of the PODs across PODs rescheduling).
* The NodePort service exposes an application service outside the Kubernetes cluster. 
HPE Container Platform automatically maps the NodePort Service endpoint to the HPE Container Platform gateway (proxy) host.

#### -4- Get the application service endpoints
To get a report on all services related to a specific virtual cluster, you can use a form of "kubectl describe" that matches against a value of the **kubedirector.hpe.com/kdcluster=YourClusterApplicationName** label.

In [23]:
tfappURL=$(kubectl describe service -l  kubedirector.hpe.com/kdcluster=${clusterName} | grep gateway1 | awk '{print $3}')
mytfapp_endpoint="https://${tfappURL}"
echo "Your application virtual cluster service endpoint is: "$mytfapp_endpoint

Your application virtual cluster service endpoint is: https://gateway1.etc.fr.comm.hpecorp.net:10002


Access to application services running in containers is proxied via the gateway host and a port number greater than 10000.

#### -5- Verify the application service is responding

In [24]:
curl -k -L -s -i "${mytfapp_endpoint}" | grep "HTTP/1.1"

HTTP/1.1 302 Found
HTTP/1.1 302 Found
HTTP/1.1 200 OK


#### -6- Fetch the authentication token of TensorFlow Jupyter Notebook 

Execute this command to get the authentication token: `kubectl exec pod-name jupyter notebook list` and use this token in Jupyter notebook:

In [25]:
mytfapp_pod=$(kubectl get pod -l kubedirector.hpe.com/kdcluster=${clusterName} | grep kdss | awk '{print $1}') 
#echo $mytfapp_pod
kubectl exec ${mytfapp_pod} jupyter notebook list

Currently running servers:
http://0.0.0.0:8888/?token=dc82e1dfbcdf9cf2df06ab44b59ac8b4201a970b68cdff47 :: /tf


#### -7- Connect to your application

You can connect to your tensorflow framework with Jupyter Notebook and start build your ML model (for example a regression model), train the model and make prediction.

Open a new tab in your browser, and connect to the service endpoint: https://77.158.163.130:YourPortNumber?token=YourToken
where 77.158.163.130 is the NAT IP address of the HPE CP proxy gateway, port number is the port you get for your service endpoint, and the token is the authentication token you have just fetched. 

![TensorFlow Jupyter Notebook](Pictures/kubedirector-tf-jupyterNotebook.png)

#### -8- Resizing the KubeDirector App cluster instance
To increase or decrease the number of members in a role, you would just have to edit the configuration YAML file and use the kubectl apply API call to implement the changes. The KubeDirector operator will manage the application cluster expansion or shrinkage for you. 

#### -9- Go through some cleanup

* First delete the kubedirector application cluster instance using the K8s API call below. The K8s API call requires the **kubectl** operation type (delete) and the application manifest.

> Note: Deleting the KubeDirectorCluster resource will automatically delete all resources (for example the pods, services, statefulset) that compose the KubeDirector application virtual cluster.

In [26]:
kubectl delete -f $tensorFlowApp

kubedirectorcluster.kubedirector.hpe.com "ml-jupyter-notebook-demouser" deleted


* Although session have a time to live (TTL) of 24 hours, it is best practice in REST API programming to cleanup and delete those sessions when done. We can use a DELETE /api/v2/session/SessionId to achieve this.

In [27]:
curl -k -i -s --request DELETE "https://${controller_endpoint}/api/v2/session/${SessionId}" \
--header "X-BDS-SESSION: $sessionlocation" \
--header 'Accept: application/json' \
--header 'Content-Type: application/json'

HTTP/1.1 204 No Content
Access-Control-Allow-Origin: *
Content-Length: 0
Content-Type: text/plain
Date: Sat, 09 May 2020 16:39:13 GMT
Server: BlueData EPIC 5.0



The status *204 No Content* means the session object has been deleted.

* Finally, reset your applications YAML files and delete the kubeconfig file

In [28]:
#reset the application deployment name in the YAML file and delete the kubeconfgi file
rm ${student}_kubeconfig
sed -i "s/${username}/example/g" $helloWorldApp
sed -i "s/${username}/example/g" $tensorFlowApp
cat $helloWorldApp
cat $tensorFlowApp

apiVersion: apps/v1
kind: Deployment
metadata:
  name: hello-world-example
spec:
  selector:
    matchLabels:
      run: hello-example
  replicas: 2
  template:
    metadata:
      labels:
        run: hello-example
    spec:
      containers:
        - name: hello-world-example
          image: gcr.io/google-samples/node-hello:1.0
          ports:
            - containerPort: 8080
              protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  name: hello-world-service-example
spec:
  selector:
    run: hello-example
  ports:
  - name: http-hello
    protocol: TCP
    port: 8080
    targetPort: 8080
  type: NodePort
---
apiVersion: "kubedirector.hpe.com/v1beta1"
kind: "KubeDirectorCluster"
metadata: 
  name: "ml-jupyter-notebook-example"
spec: 
  app: "ml-jupyter-notebook"
  appCatalog: "local"
  roles: 
    - 
      id: "controller"
      members: 1
      resources: 
        requests: 
          memory: "4Gi"
          cpu: "2"
          nvidia.com/gpu: "0"
        limits: 


## Summary

In this tutorial, you learned how **tenant users** can deploy, programmatically, cloud-native stateless, microservices applications, and non-cloud-native distributed stateful AI/analytics KubeDirector applications on Kubernetes clusters managed by HPE Container Platform. You also used the standard Kubernetes command line (kubectl) to interact with the Kubernetes cluster in the context of your tenant user account. 